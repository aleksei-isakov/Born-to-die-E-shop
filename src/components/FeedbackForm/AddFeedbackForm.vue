<template>
  <feedback-form
    ref="addFeedbackForm"
    :feedback="newFeedback"
    :is-dialog-active="isDialogActive"
    @close="closeDialog"
    @submit="onSubmitSendFeedback"
  />
</template>

<script>
import FeedbackForm from '@/components/FeedbackForm/FeedbackForm.vue';
import { mapGetters, mapActions } from 'vuex';
var faker = require('faker');

export default {
  name: 'AddFeedbackForm',

  components: { FeedbackForm },

  props: {
    userId: {
      type: String,
      default: ''
    },

    isDialogActive: {
      type: Boolean,
      default: false
    }
  },

  data() {
    return {
      newFeedback: {
        reviewerName: null,
        comment: null,
        date: null,
        rating: 0
      }
    };
  },

  computed: {
    ...mapGetters('PdpPageModule', ['productInfo', 'isError'])
  },

  methods: {
    ...mapActions('PdpPageModule', [
      'updateProductInfoWithNewFeedback',
      'addFeedbackIntoProductInfo'
    ]),

    closeDialog() {
      this.$emit('close');
    },

    setFormCurrentDate() {
      this.newFeedback.date = Date();
    },

    onSubmitClearFormData() {
      this.$refs.addFeedbackForm.clearFormData();
    },

    async onSubmitSendFeedback() {
      this.setFormCurrentDate();
      this.addFeedbackIntoProductInfo({
        ...this.newFeedback,
        id: faker.datatype.uuid(),
        reviewerId: this.userId
      });

      const payload = {
        productId: this.$route.params.id,
        updatedProductInfo: this.productInfo
      };

      await this.updateProductInfoWithNewFeedback(payload);

      if (!this.isError) {
        this.onSubmitClearFormData();
        this.closeDialog();
      }
    }
  }
};
</script>
