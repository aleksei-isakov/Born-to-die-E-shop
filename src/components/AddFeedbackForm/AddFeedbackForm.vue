<template>
  <feedback-form
    ref="addFeedbackForm"
    :feedback="newFeedback"
    :is-dialog-active="isDialogActive"
    @close="closeDialog"
    @submit="onSubmitSendFeedback"
  />
</template>

<script>
import FeedbackForm from '@/components/FeedbackForm/FeedbackForm.vue';
import { mapGetters, mapActions } from 'vuex';
var faker = require('faker');

export default {
  name: 'CreateFeedbackForm',

  components: { FeedbackForm },

  props: {
    userId: {
      type: String,
      default: ''
    },

    isDialogActive: {
      type: Boolean,
      default: false
    }
  },

  data() {
    return {
      newFeedback: {
        reviewerName: null,
        comment: null,
        date: null,
        rating: 0,
        reviewerId: this.userId,
        id: faker.datatype.uuid()
      }
    };
  },

  computed: {
    ...mapGetters('PdpPageModule', ['productInfo', 'isError'])
  },

  methods: {
    ...mapActions('PdpPageModule', [
      'updateProductInfoWithNewFeedback',
      'addFeedbackIntoProductInfo'
    ]),

    closeDialog() {
      this.$emit('close');
    },

    setFormCurrentDate() {
      this.newFeedback.date = `${new Date()}`;
    },

    onSubmitClearFormData() {
      this.$refs.addFeedbackForm.clearFormData();
    },

    async onSubmitSendFeedback() {
      this.setFormCurrentDate();
      this.addFeedbackIntoProductInfo({ ...this.newFeedback });

      const payload = {
        productId: this.$route.params.id,
        updatedProductInfo: this.productInfo
      };

      await this.updateProductInfoWithNewFeedback(payload);

      if (!this.isError) {
        this.onSubmitClearFormData();
        this.closeDialog();
      }
    }
  }
};
</script>
