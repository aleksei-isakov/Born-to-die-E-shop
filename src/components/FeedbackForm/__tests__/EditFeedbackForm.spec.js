import { mount, createLocalVue } from '@vue/test-utils';
import AddFeedbackForm from '@/components/FeedbackForm/AddFeedbackForm';
import Vuex from 'vuex';
import feedbacks from '@/components/Feedback/mocks.json';
import Vuetify from 'vuetify';
import Vue from 'vue';
import VueRouter from 'vue-router';

describe('AddFeedbackForm.vue', () => {
  let wrapper;
  let getters;
  let actions;
  let vuetify;
  let store;
  let router;
  const localVue = createLocalVue();

  Vue.use(Vuetify);
  localVue.use(VueRouter);
  localVue.use(Vuex);

  const mockProductInfo = {
    id: '6f40cf0f-9841-4994-ae04-2203145592a5',
    name: 'Fantastic Soft Sausages',
    description:
      'The slim & simple Maple Gaming Keyboard from Dev Byte comes with a sleek body and 7- Color RGB LED Back-lighting for smart functionality',
    price: 852,
    images: [
      'http://placeimg.com/640/480',
      'http://placeimg.com/640/480',
      'http://placeimg.com/640/480'
    ],
    createdAt: '2022-01-31T09:22:07.577Z',
    updatedAt: '',
    feedbacks: feedbacks['feedbacks']
  };

  beforeEach(() => {
    getters = {
      productInfo: () => {}
    };

    actions = {
      updateProductInfoWithNewFeedback: jest.fn(
        (getters.productInfo = () => mockProductInfo)
      ),
      addFeedbackIntoProductInfo: jest.fn(
        (getters.productInfo = () => mockProductInfo)
      )
    };

    vuetify = new Vuetify();
    router = new VueRouter();

    store = new Vuex.Store({
      modules: {
        PdpPageModule: {
          getters,
          actions
        }
      }
    });

    wrapper = mount(AddFeedbackForm, {
      localVue,
      vuetify,
      store,
      router,
      propsData: {
        isDialogActive: true
      }
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  it('should render form for editing feedback correct', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should open filled feedback form', async () => {
    const reviewerName = wrapper.get('#feedback-v-text-field');
    const rating = wrapper.get('#rating-5');
    const comment = wrapper.get('#feedback-v-textarea');
    const openForm = async () => {
      await reviewerName.setValue('Lorem');
      await rating.setChecked();
      await comment.setValue('Lorem ipsum dolor sit amet, consectetur.');
    };

    await openForm();
    expect(reviewerName.element.value).toBe('Lorem');
    expect(rating.element.value).toBe('5');
    expect(comment.element.value).toBe(
      'Lorem ipsum dolor sit amet, consectetur.'
    );
  });
});
