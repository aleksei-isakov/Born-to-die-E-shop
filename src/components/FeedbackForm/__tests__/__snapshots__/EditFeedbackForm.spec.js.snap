// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddFeedbackForm.vue should render form for adding feedback correct 1`] = `
<div
  class="v-dialog__container"
>
  <div
    aria-modal="true"
    class="v-dialog__content v-dialog__content--active"
    role="dialog"
    style="z-index: 202;"
    tabindex="0"
  >
    <transition-stub
      appear="true"
      name="dialog-transition"
    >
      <div
        class="v-dialog v-dialog--active"
        style="transform-origin: center center; width: 600px;"
      >
        <form
          class="v-form feedback-form"
          novalidate="novalidate"
        >
          <div
            class="v-card v-sheet theme--light"
          >
            <div
              class="container feedback__container"
            >
              <div
                class="feedback__icon"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate fas fa-times theme--light blue--text"
                />
              </div>
               
              <div
                class="row"
              >
                <div
                  class="col col-12"
                >
                  <div
                    class="v-input feedback-form__name theme--light v-text-field"
                  >
                    <div
                      class="v-input__control"
                    >
                      <div
                        class="v-input__slot"
                      >
                        <div
                          class="v-text-field__slot"
                        >
                          <label
                            class="v-label theme--light"
                            for="feedback-v-text-field"
                            style="left: 0px; position: absolute;"
                          >
                            Enter your name
                          </label>
                          <input
                            id="feedback-v-text-field"
                            type="text"
                          />
                        </div>
                      </div>
                      <div
                        class="v-text-field__details"
                      >
                        <div
                          class="v-messages theme--light"
                        >
                          <transition-group-stub
                            class="v-messages__wrapper"
                            name="message-transition"
                            tag="div"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                 
                <div
                  align="start"
                  class="col col-12"
                >
                  <p
                    class="rating__title"
                  >
                    Rating
                  </p>
                   
                  <div
                    class="feedback-form__rating"
                  >
                    <div>
                      <div
                        class="rating-group"
                      >
                        <div
                          class="stars-wrapper"
                        >
                          <label
                            aria-label="1 stars"
                            class="rating__label"
                            for="rating-1"
                          >
                            <i
                              class="rating__icon fa fa-star"
                            />
                          </label>
                           
                          <input
                            class="rating__input"
                            id="rating-1"
                            name="rating"
                            type="radio"
                            value="1"
                          />
                        </div>
                        <div
                          class="stars-wrapper"
                        >
                          <label
                            aria-label="2 stars"
                            class="rating__label"
                            for="rating-2"
                          >
                            <i
                              class="rating__icon fa fa-star"
                            />
                          </label>
                           
                          <input
                            class="rating__input"
                            id="rating-2"
                            name="rating"
                            type="radio"
                            value="2"
                          />
                        </div>
                        <div
                          class="stars-wrapper"
                        >
                          <label
                            aria-label="3 stars"
                            class="rating__label"
                            for="rating-3"
                          >
                            <i
                              class="rating__icon fa fa-star"
                            />
                          </label>
                           
                          <input
                            class="rating__input"
                            id="rating-3"
                            name="rating"
                            type="radio"
                            value="3"
                          />
                        </div>
                        <div
                          class="stars-wrapper"
                        >
                          <label
                            aria-label="4 stars"
                            class="rating__label"
                            for="rating-4"
                          >
                            <i
                              class="rating__icon fa fa-star"
                            />
                          </label>
                           
                          <input
                            class="rating__input"
                            id="rating-4"
                            name="rating"
                            type="radio"
                            value="4"
                          />
                        </div>
                        <div
                          class="stars-wrapper"
                        >
                          <label
                            aria-label="5 stars"
                            class="rating__label"
                            for="rating-5"
                          >
                            <i
                              class="rating__icon fa fa-star"
                            />
                          </label>
                           
                          <input
                            class="rating__input"
                            id="rating-5"
                            name="rating"
                            type="radio"
                            value="5"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                 
                <div
                  class="col col-12"
                >
                  <div
                    class="v-input feedback-form__comment v-textarea v-textarea--auto-grow v-textarea--no-resize theme--light v-text-field v-text-field--enclosed v-text-field--outlined"
                  >
                    <div
                      class="v-input__control"
                    >
                      <div
                        class="v-input__slot"
                      >
                        <fieldset
                          aria-hidden="true"
                        >
                          <legend
                            style="width: 0px;"
                          >
                            <span
                              class="notranslate"
                            >
                              â€‹
                            </span>
                          </legend>
                        </fieldset>
                        <div
                          class="v-text-field__slot"
                        >
                          <label
                            class="v-label theme--light"
                            for="feedback-v-textarea"
                            style="left: 0px; position: absolute;"
                          >
                            Comment
                          </label>
                          <textarea
                            id="feedback-v-textarea"
                            rows="1"
                          />
                        </div>
                      </div>
                      <div
                        class="v-text-field__details"
                      >
                        <div
                          class="v-messages theme--light"
                        >
                          <transition-group-stub
                            class="v-messages__wrapper"
                            name="message-transition"
                            tag="div"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                 
                <div
                  class="col col-12"
                >
                  <button
                    class="base-button base-text-button base-text-filled-button feedback__btn custom-blue"
                    disabled="disabled"
                  >
                    
              ADD NEW FEEDBACK
            
                  </button>
                </div>
              </div>
            </div>
          </div>
        </form>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`AddFeedbackForm.vue should render form for editing feedback correct 1`] = `
<div
  class="v-dialog__container"
>
  <div
    aria-modal="true"
    class="v-dialog__content v-dialog__content--active"
    role="dialog"
    style="z-index: 202;"
    tabindex="0"
  >
    <transition-stub
      appear="true"
      name="dialog-transition"
    >
      <div
        class="v-dialog v-dialog--active"
        style="transform-origin: center center; width: 600px;"
      >
        <form
          class="v-form feedback-form"
          novalidate="novalidate"
        >
          <div
            class="v-card v-sheet theme--light"
          >
            <div
              class="container feedback__container"
            >
              <div
                class="feedback__icon"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate fas fa-times theme--light blue--text"
                />
              </div>
               
              <div
                class="row"
              >
                <div
                  class="col col-12"
                >
                  <div
                    class="v-input feedback-form__name theme--light v-text-field"
                  >
                    <div
                      class="v-input__control"
                    >
                      <div
                        class="v-input__slot"
                      >
                        <div
                          class="v-text-field__slot"
                        >
                          <label
                            class="v-label theme--light"
                            for="feedback-v-text-field"
                            style="left: 0px; position: absolute;"
                          >
                            Enter your name
                          </label>
                          <input
                            id="feedback-v-text-field"
                            type="text"
                          />
                        </div>
                      </div>
                      <div
                        class="v-text-field__details"
                      >
                        <div
                          class="v-messages theme--light"
                        >
                          <transition-group-stub
                            class="v-messages__wrapper"
                            name="message-transition"
                            tag="div"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                 
                <div
                  align="start"
                  class="col col-12"
                >
                  <p
                    class="rating__title"
                  >
                    Rating
                  </p>
                   
                  <div
                    class="feedback-form__rating"
                  >
                    <div>
                      <div
                        class="rating-group"
                      >
                        <div
                          class="stars-wrapper"
                        >
                          <label
                            aria-label="1 stars"
                            class="rating__label"
                            for="rating-1"
                          >
                            <i
                              class="rating__icon fa fa-star"
                            />
                          </label>
                           
                          <input
                            class="rating__input"
                            id="rating-1"
                            name="rating"
                            type="radio"
                            value="1"
                          />
                        </div>
                        <div
                          class="stars-wrapper"
                        >
                          <label
                            aria-label="2 stars"
                            class="rating__label"
                            for="rating-2"
                          >
                            <i
                              class="rating__icon fa fa-star"
                            />
                          </label>
                           
                          <input
                            class="rating__input"
                            id="rating-2"
                            name="rating"
                            type="radio"
                            value="2"
                          />
                        </div>
                        <div
                          class="stars-wrapper"
                        >
                          <label
                            aria-label="3 stars"
                            class="rating__label"
                            for="rating-3"
                          >
                            <i
                              class="rating__icon fa fa-star"
                            />
                          </label>
                           
                          <input
                            class="rating__input"
                            id="rating-3"
                            name="rating"
                            type="radio"
                            value="3"
                          />
                        </div>
                        <div
                          class="stars-wrapper"
                        >
                          <label
                            aria-label="4 stars"
                            class="rating__label"
                            for="rating-4"
                          >
                            <i
                              class="rating__icon fa fa-star"
                            />
                          </label>
                           
                          <input
                            class="rating__input"
                            id="rating-4"
                            name="rating"
                            type="radio"
                            value="4"
                          />
                        </div>
                        <div
                          class="stars-wrapper"
                        >
                          <label
                            aria-label="5 stars"
                            class="rating__label"
                            for="rating-5"
                          >
                            <i
                              class="rating__icon fa fa-star"
                            />
                          </label>
                           
                          <input
                            class="rating__input"
                            id="rating-5"
                            name="rating"
                            type="radio"
                            value="5"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                 
                <div
                  class="col col-12"
                >
                  <div
                    class="v-input feedback-form__comment v-textarea v-textarea--auto-grow v-textarea--no-resize theme--light v-text-field v-text-field--enclosed v-text-field--outlined"
                  >
                    <div
                      class="v-input__control"
                    >
                      <div
                        class="v-input__slot"
                      >
                        <fieldset
                          aria-hidden="true"
                        >
                          <legend
                            style="width: 0px;"
                          >
                            <span
                              class="notranslate"
                            >
                              â€‹
                            </span>
                          </legend>
                        </fieldset>
                        <div
                          class="v-text-field__slot"
                        >
                          <label
                            class="v-label theme--light"
                            for="feedback-v-textarea"
                            style="left: 0px; position: absolute;"
                          >
                            Comment
                          </label>
                          <textarea
                            id="feedback-v-textarea"
                            rows="1"
                          />
                        </div>
                      </div>
                      <div
                        class="v-text-field__details"
                      >
                        <div
                          class="v-messages theme--light"
                        >
                          <transition-group-stub
                            class="v-messages__wrapper"
                            name="message-transition"
                            tag="div"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                 
                <div
                  class="col col-12"
                >
                  <button
                    class="base-button base-text-button base-text-filled-button feedback__btn custom-blue"
                    disabled="disabled"
                  >
                    
              ADD NEW FEEDBACK
            
                  </button>
                </div>
              </div>
            </div>
          </div>
        </form>
      </div>
    </transition-stub>
  </div>
</div>
`;
